File: engine/dome.py
"""
Kuppelplättchen-Definitionen für Mosaic-AI.

Jede Kuppelplatte ist eine 2×2-Platte. Die Spieler legen sie auf ihr 6×6-Raster
(3×3-Anordnung von Kuppelplättchen).

Space-Typen laut dome_color.csv:
  - NORMAL(farbe): nur diese Farbe darf hier platziert werden
  - WILD ("bunt"): jede Farbe darf hier platziert werden
  - SPECIAL ("weiß"): wird erst freigeschaltet wenn die anderen 3 Spaces
                      der Kuppel gefüllt sind; nimmt einen weißen Stein
                      aus dem separaten Vorrat auf

Die 18 Kuppeln stammen direkt aus dome_color.csv.
Reihenfolge der 4 Spaces pro Kuppel: [oben-links, oben-rechts, unten-links, unten-rechts]
"""

from __future__ import annotations
from dataclasses import dataclass, field
from enum import Enum, auto
from typing import Optional

from engine.tile import TileColor


class SpaceType(Enum):
    NORMAL  = auto()   # nimmt genau eine Farbe an
    WILD    = auto()   # nimmt jede Farbe an
    SPECIAL = auto()   # weißer Stein, erst nach Unlock verfügbar


@dataclass
class DomeSpace:
    """
    Einer der 4 Spaces auf einer Kuppelplättchen.

    Normal/Wild-Spaces speichern welche TileColor platziert wurde.
    Special-Spaces (weiß) speichern nur ob sie befüllt wurden —
    weiße Steine haben keine Farbe.
    """
    space_type:     SpaceType
    required_color: Optional[TileColor] = None  # nur bei NORMAL gesetzt
    placed_color:   Optional[TileColor] = None  # befüllt bei NORMAL / WILD
    placed_special: bool                = False  # befüllt bei SPECIAL
    is_locked:      bool                = False  # True bei SPECIAL bis Unlock

    @property
    def is_filled(self) -> bool:
        if self.space_type == SpaceType.SPECIAL:
            return self.placed_special
        return self.placed_color is not None

    def accepts(self, color: TileColor) -> bool:
        """Kann dieser Space einen normalen Stein dieser Farbe aufnehmen?"""
        if self.is_filled or self.is_locked:
            return False
        if self.space_type == SpaceType.NORMAL:
            return color == self.required_color
        if self.space_type == SpaceType.WILD:
            return True
        return False  # SPECIAL nur via place_special_tile()

    def accepts_special(self) -> bool:
        """Kann dieser Space einen weißen Stein aufnehmen?"""
        return (
            self.space_type == SpaceType.SPECIAL
            and not self.is_locked
            and not self.placed_special
        )

    def place_special_tile(self) -> None:
        if not self.accepts_special():
            raise ValueError("Dieser Space kann keinen weißen Stein aufnehmen.")
        self.placed_special = True

    def __repr__(self) -> str:
        if self.space_type == SpaceType.NORMAL:
            filled = self.placed_color.value if self.placed_color else "leer"
            return f"Normal({self.required_color.value}, {filled})"
        if self.space_type == SpaceType.WILD:
            filled = self.placed_color.value if self.placed_color else "leer"
            return f"Wild({filled})"
        state = "gefüllt" if self.placed_special else ("offen" if not self.is_locked else "gesperrt")
        return f"Special({state})"
        
    def clone(self) -> "DomeSpace":
        new_space = self.__class__(self.space_type) # Nimmt den Typ (WILD, NORMAL, SPECIAL)
        new_space.placed_color = self.placed_color
        new_space.is_locked = self.is_locked
        new_space.placed_special = getattr(self, 'placed_special', False)
        return new_space    


# Rotations-Mapping: Rotation → neue Reihenfolge der Space-Indizes
# Layout vor Rotation:  [0][1]
#                       [2][3]
# 90° im Uhrzeigersinn: pos0←pos2, pos1←pos0, pos2←pos3, pos3←pos1
ROTATION_MAP: dict[int, list[int]] = {
      0: [0, 1, 2, 3],   # original
     90: [2, 0, 3, 1],   # 90° im Uhrzeigersinn
    180: [3, 2, 1, 0],   # 180°
    270: [1, 3, 0, 2],   # 270° im Uhrzeigersinn
}


@dataclass
class DomeTile:
    """
    Eine 2×2 Kuppelplättchen.

    spaces: 4 DomeSpace-Objekte, Index-Layout:
        [0] [1]   →   oben-links   oben-rechts
        [2] [3]   →   unten-links  unten-rechts

    bonus_points: Zusatzpunkte wenn der SPECIAL-Space befüllt wird
                  (0 wenn kein SPECIAL-Space vorhanden)
    tile_id: 0-basierter Index in der Kuppel-Tabelle (dome_color.csv)
    """
    tile_id:      int
    spaces:       list[DomeSpace]
    bonus_points: int = 0

    def __post_init__(self):
        assert len(self.spaces) == 4, "Eine Kuppelplättchen hat genau 4 Spaces"

    @property
    def is_complete(self) -> bool:
        return all(s.is_filled for s in self.spaces)

    @property
    def special_space(self) -> Optional[DomeSpace]:
        for s in self.spaces:
            if s.space_type == SpaceType.SPECIAL:
                return s
        return None

    def try_unlock_special(self) -> bool:
        """
        Prüft ob der SPECIAL-Space freigeschaltet werden soll
        (sobald die anderen 3 Spaces gefüllt sind).
        Gibt True zurück wenn gerade freigeschaltet wurde.
        """
        sp = self.special_space
        if sp is None or not sp.is_locked:
            return False
        other_filled = all(s.is_filled for s in self.spaces if s is not sp)
        if other_filled:
            sp.is_locked = False
            return True
        return False

    # ------------------------------------------------------------------
    # Rotation
    # ------------------------------------------------------------------

    def rotated_spaces(self, degrees: int) -> list[DomeSpace]:
        """
        Gibt die 4 Spaces in der gedrehten Reihenfolge zurück.
        degrees muss 0, 90, 180 oder 270 sein.
        Ändert die Kuppel selbst NICHT — nur für Vorschau/Validierung.
        """
        if degrees not in ROTATION_MAP:
            raise ValueError(f"Ungültige Rotation: {degrees}. Erlaubt: 0, 90, 180, 270.")
        indices = ROTATION_MAP[degrees]
        return [self.spaces[i] for i in indices]

    def apply_rotation(self, degrees: int) -> None:
        """
        Dreht die Kuppel dauerhaft um degrees Grad.
        Darf nur VOR dem Platzieren auf dem Brett aufgerufen werden.
        """
        if degrees not in ROTATION_MAP:
            raise ValueError(f"Ungültige Rotation: {degrees}. Erlaubt: 0, 90, 180, 270.")
        if degrees == 0:
            return
        if any(s.is_filled for s in self.spaces):
            raise ValueError("Eine bereits befüllte Kuppel kann nicht rotiert werden.")
        self.spaces = self.rotated_spaces(degrees)

    def open_spaces_for(self, color: TileColor) -> list[int]:
        """Indizes der Spaces die diese Farbe aufnehmen können."""
        return [i for i, s in enumerate(self.spaces) if s.accepts(color)]

    def __repr__(self) -> str:
        return (
            f"DomeTile(id={self.tile_id}, "
            f"spaces={self.spaces}, bonus={self.bonus_points})"
        )

    def clone(self) -> "DomeTile":
        # 1. Zuerst die 4 Spaces einzeln klonen
        cloned_spaces = [space.clone() for space in self.spaces]

        # 2. Das neue Tile direkt mit den geklonten Spaces initialisieren
        # (Das verhindert, dass __post_init__ sich über fehlende Spaces beschwert)
        new_tile = DomeTile(
            tile_id=self.tile_id,
            spaces=cloned_spaces
        )
        return new_tile

# ---------------------------------------------------------------------------
# Hilfsfunktionen
# ---------------------------------------------------------------------------

def _n(color: TileColor) -> DomeSpace:
    """Normal-Space für gegebene Farbe."""
    return DomeSpace(SpaceType.NORMAL, required_color=color)

def _w() -> DomeSpace:
    """Wild-Space (bunt)."""
    return DomeSpace(SpaceType.WILD)

def _s() -> DomeSpace:
    """Special-Space (weiß), startet gesperrt."""
    return DomeSpace(SpaceType.SPECIAL, is_locked=True)


# ---------------------------------------------------------------------------
# Die 18 Kuppelplättchen aus dome_colors.csv
# Reihenfolge der Spalten: oben-links, oben-rechts, unten-links, unten-rechts
# "weiß" → _s()  |  "bunt" → _w()  |  alles andere → _n(farbe)
# ---------------------------------------------------------------------------

def build_dome_tile_pool() -> list[DomeTile]:
    """
    Gibt den vollständigen Pool von 18 Kuppelplättchen zurück.
    Direkt aus dome_colors.csv übernommen.
    """
    B  = TileColor.BLAU
    G  = TileColor.GELB
    R  = TileColor.ROT
    S  = TileColor.SCHWARZ
    T  = TileColor.TUERKIS

    # Jede Zeile: (spaces_liste, bonus_points)
    # bonus_points für Special-Spaces: je nach Schwierigkeit 3-5 Pkt
    # (weiße Spaces = _s(), bunt = _w())
    definitions: list[tuple[list[DomeSpace], int]] = [
        # Zeile  1: gelb    schwarz  türkis  weiß
        ([_n(G), _n(S), _n(T), _s()], 3),
        # Zeile  2: bunt    blau     türkis  schwarz
        ([_w(),  _n(B), _n(T), _n(S)], 0),
        # Zeile  3: türkis  rot      blau    bunt
        ([_n(T), _n(R), _n(B), _w()], 0),
        # Zeile  4: schwarz gelb     rot     bunt
        ([_n(S), _n(G), _n(R), _w()], 0),
        # Zeile  5: schwarz weiß     türkis  rot
        ([_n(S), _s(),  _n(T), _n(R)], 3),
        # Zeile  6: türkis  gelb     bunt    schwarz
        ([_n(T), _n(G), _w(),  _n(S)], 0),
        # Zeile  7: weiß    schwarz  rot     blau
        ([_s(),  _n(S), _n(R), _n(B)], 3),
        # Zeile  8: gelb    blau     schwarz weiß
        ([_n(G), _n(B), _n(S), _s()], 3),
        # Zeile  9: türkis  rot      blau    weiß
        ([_n(T), _n(R), _n(B), _s()], 3),
        # Zeile 10: gelb    rot      bunt    blau
        ([_n(G), _n(R), _w(),  _n(B)], 0),
        # Zeile 11: gelb    weiß     schwarz rot
        ([_n(G), _s(),  _n(S), _n(R)], 3),
        # Zeile 12: türkis  schwarz  rot     bunt
        ([_n(T), _n(S), _n(R), _w()], 0),
        # Zeile 13: blau    schwarz  weiß    türkis
        ([_n(B), _n(S), _s(),  _n(T)], 3),
        # Zeile 14: rot     türkis   gelb    bunt
        ([_n(R), _n(T), _n(G), _w()], 0),
        # Zeile 15: türkis  blau     bunt    gelb
        ([_n(T), _n(B), _w(),  _n(G)], 0),
        # Zeile 16: weiß    türkis   gelb    blau
        ([_s(),  _n(T), _n(G), _n(B)], 3),
        # Zeile 17: rot     bunt     blau    schwarz
        ([_n(R), _w(),  _n(B), _n(S)], 0),
        # Zeile 18: weiß    gelb     blau    rot
        ([_s(),  _n(G), _n(B), _n(R)], 3),
    ]

    return [
        DomeTile(tile_id=i, spaces=spaces, bonus_points=bp)
        for i, (spaces, bp) in enumerate(definitions)
    ]


# ---------------------------------------------------------------------------
# Die 20 Bonusplättchen aus bonus_chips_colors.csv
# Jedes Plättchen hat 1 oder 2 Farbfelder.
# ---------------------------------------------------------------------------

@dataclass
class BonusChip:
    """
    Ein Bonusplättchen.
    colors: 1 oder 2 Farben die dieses Plättchen zeigt.
    chip_id: 0-basierter Index.
    """
    chip_id: int
    colors:  list[TileColor]

    def __repr__(self) -> str:
        return f"BonusChip({[c.value for c in self.colors]})"


def build_bonus_chip_pool() -> list[BonusChip]:
    """
    Gibt alle 20 Bonusplättchen zurück.
    Direkt aus bonus_chips_colors.csv
    """
    B = TileColor.BLAU
    G = TileColor.GELB
    R = TileColor.ROT
    S = TileColor.SCHWARZ
    T = TileColor.TUERKIS

    definitions: list[list[TileColor]] = [
        [B],        # Blau
        [T],        # Türkis
        [T, G],     # Türkis + gelb
        [B, R],     # Blau + rot
        [R],        # Rot
        [R, T],     # rot + türkis
        [S, B],     # schwarz + blau
        [G, S],     # gelb + schwarz
        [B],        # Blau
        [R],        # rot
        [B, R],     # blau + rot
        [S],        # schwarz
        [S, G],     # schwarz + gelb
        [S],        # schwarz
        [R, T],     # rot + türkis
        [B, S],     # Blau + schwarz
        [G],        # gelb
        [T, G],     # Türkis + gelb
        [T],        # Türkis
        [G],        # gelb
    ]

    return [
        BonusChip(chip_id=i, colors=colors)
        for i, colors in enumerate(definitions)
    ]

File: rust/src/dome.rs
//! Kuppelplättchen (2×2) und Bonusplättchen — Port von engine/dome.py.

use crate::tile::TileColor;

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum SpaceType {
    Normal,
    Wild,
    Special,
}

/// Einer der 4 Spaces auf einer Kuppelplättchen.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct DomeSpace {
    pub space_type: SpaceType,
    pub required_color: Option<TileColor>,
    pub placed_color: Option<TileColor>,
    pub placed_special: bool,
    pub is_locked: bool,
}

impl DomeSpace {
    pub fn normal(color: TileColor) -> Self {
        DomeSpace {
            space_type: SpaceType::Normal,
            required_color: Some(color),
            placed_color: None,
            placed_special: false,
            is_locked: false,
        }
    }

    pub fn wild() -> Self {
        DomeSpace {
            space_type: SpaceType::Wild,
            required_color: None,
            placed_color: None,
            placed_special: false,
            is_locked: false,
        }
    }

    /// Special-Space (weiß), startet gesperrt.
    pub fn special() -> Self {
        DomeSpace {
            space_type: SpaceType::Special,
            required_color: None,
            placed_color: None,
            placed_special: false,
            is_locked: true,
        }
    }

    pub fn is_filled(&self) -> bool {
        match self.space_type {
            SpaceType::Special => self.placed_special,
            _ => self.placed_color.is_some(),
        }
    }

    /// Kann dieser Space einen normalen Stein dieser Farbe aufnehmen?
    pub fn accepts(&self, color: TileColor) -> bool {
        if self.is_filled() || self.is_locked {
            return false;
        }
        match self.space_type {
            SpaceType::Normal => Some(color) == self.required_color,
            SpaceType::Wild => true,
            SpaceType::Special => false,
        }
    }

    /// Kann dieser Space einen weißen Stein aufnehmen?
    pub fn accepts_special(&self) -> bool {
        self.space_type == SpaceType::Special && !self.is_locked && !self.placed_special
    }

    pub fn place_special_tile(&mut self) -> Result<(), String> {
        if !self.accepts_special() {
            return Err("Dieser Space kann keinen weißen Stein aufnehmen.".into());
        }
        self.placed_special = true;
        Ok(())
    }
}

/// Rotation → neue Reihenfolge der Space-Indizes.
/// Layout vor Rotation: [0][1] / [2][3].
pub fn rotation_indices(degrees: u32) -> Option<[usize; 4]> {
    match degrees {
        0 => Some([0, 1, 2, 3]),
        90 => Some([2, 0, 3, 1]),
        180 => Some([3, 2, 1, 0]),
        270 => Some([1, 3, 0, 2]),
        _ => None,
    }
}

/// Eine 2×2 Kuppelplättchen. spaces-Index-Layout: [0][1] / [2][3].
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct DomeTile {
    pub tile_id: usize,
    pub spaces: Vec<DomeSpace>, // genau 4
    pub bonus_points: i32,
}

impl DomeTile {
    pub fn new(tile_id: usize, spaces: Vec<DomeSpace>, bonus_points: i32) -> Self {
        assert_eq!(spaces.len(), 4, "Eine Kuppelplättchen hat genau 4 Spaces");
        DomeTile {
            tile_id,
            spaces,
            bonus_points,
        }
    }

    pub fn is_complete(&self) -> bool {
        self.spaces.iter().all(|s| s.is_filled())
    }

    /// Index des SPECIAL-Space, falls vorhanden.
    pub fn special_space_idx(&self) -> Option<usize> {
        self.spaces
            .iter()
            .position(|s| s.space_type == SpaceType::Special)
    }

    /// Schaltet den SPECIAL-Space frei, sobald die anderen 3 gefüllt sind.
    /// Gibt true zurück, wenn gerade freigeschaltet wurde.
    pub fn try_unlock_special(&mut self) -> bool {
        let sp_idx = match self.special_space_idx() {
            Some(i) => i,
            None => return false,
        };
        if !self.spaces[sp_idx].is_locked {
            return false;
        }
        let other_filled = self
            .spaces
            .iter()
            .enumerate()
            .all(|(i, s)| i == sp_idx || s.is_filled());
        if other_filled {
            self.spaces[sp_idx].is_locked = false;
            return true;
        }
        false
    }

    /// Spaces in gedrehter Reihenfolge (ändert die Platte nicht).
    pub fn rotated_spaces(&self, degrees: u32) -> Result<Vec<DomeSpace>, String> {
        let idx = rotation_indices(degrees)
            .ok_or_else(|| format!("Ungültige Rotation: {degrees}. Erlaubt: 0, 90, 180, 270."))?;
        Ok(idx.iter().map(|&i| self.spaces[i].clone()).collect())
    }

    /// Dreht die Platte dauerhaft (nur vor dem Platzieren erlaubt).
    pub fn apply_rotation(&mut self, degrees: u32) -> Result<(), String> {
        if rotation_indices(degrees).is_none() {
            return Err(format!(
                "Ungültige Rotation: {degrees}. Erlaubt: 0, 90, 180, 270."
            ));
        }
        if degrees == 0 {
            return Ok(());
        }
        if self.spaces.iter().any(|s| s.is_filled()) {
            return Err("Eine bereits befüllte Kuppel kann nicht rotiert werden.".into());
        }
        self.spaces = self.rotated_spaces(degrees)?;
        Ok(())
    }

    /// Indizes der Spaces, die diese Farbe aufnehmen können.
    pub fn open_spaces_for(&self, color: TileColor) -> Vec<usize> {
        self.spaces
            .iter()
            .enumerate()
            .filter(|(_, s)| s.accepts(color))
            .map(|(i, _)| i)
            .collect()
    }
}

/// Vollständiger Pool von 18 Kuppelplättchen (engine/dome.py, dome_colors.csv).
pub fn build_dome_tile_pool() -> Vec<DomeTile> {
    use TileColor::*;
    let n = DomeSpace::normal;
    let w = DomeSpace::wild;
    let s = DomeSpace::special;

    // (spaces, bonus_points); Reihenfolge: oben-links, oben-rechts, unten-links, unten-rechts
    let defs: Vec<(Vec<DomeSpace>, i32)> = vec![
        (vec![n(Gelb), n(Schwarz), n(Tuerkis), s()], 3),
        (vec![w(), n(Blau), n(Tuerkis), n(Schwarz)], 0),
        (vec![n(Tuerkis), n(Rot), n(Blau), w()], 0),
        (vec![n(Schwarz), n(Gelb), n(Rot), w()], 0),
        (vec![n(Schwarz), s(), n(Tuerkis), n(Rot)], 3),
        (vec![n(Tuerkis), n(Gelb), w(), n(Schwarz)], 0),
        (vec![s(), n(Schwarz), n(Rot), n(Blau)], 3),
        (vec![n(Gelb), n(Blau), n(Schwarz), s()], 3),
        (vec![n(Tuerkis), n(Rot), n(Blau), s()], 3),
        (vec![n(Gelb), n(Rot), w(), n(Blau)], 0),
        (vec![n(Gelb), s(), n(Schwarz), n(Rot)], 3),
        (vec![n(Tuerkis), n(Schwarz), n(Rot), w()], 0),
        (vec![n(Blau), n(Schwarz), s(), n(Tuerkis)], 3),
        (vec![n(Rot), n(Tuerkis), n(Gelb), w()], 0),
        (vec![n(Tuerkis), n(Blau), w(), n(Gelb)], 0),
        (vec![s(), n(Tuerkis), n(Gelb), n(Blau)], 3),
        (vec![n(Rot), w(), n(Blau), n(Schwarz)], 0),
        (vec![s(), n(Gelb), n(Blau), n(Rot)], 3),
    ];

    defs.into_iter()
        .enumerate()
        .map(|(i, (spaces, bp))| DomeTile::new(i, spaces, bp))
        .collect()
}

/// Bonusplättchen (1–2 Farben).
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct BonusChip {
    pub chip_id: usize,
    pub colors: Vec<TileColor>,
}

/// Alle 20 Bonusplättchen (engine/dome.py, bonus_chips_colors.csv).
pub fn build_bonus_chip_pool() -> Vec<BonusChip> {
    use TileColor::*;
    let defs: Vec<Vec<TileColor>> = vec![
        vec![Blau],
        vec![Tuerkis],
        vec![Tuerkis, Gelb],
        vec![Blau, Rot],
        vec![Rot],
        vec![Rot, Tuerkis],
        vec![Schwarz, Blau],
        vec![Gelb, Schwarz],
        vec![Blau],
        vec![Rot],
        vec![Blau, Rot],
        vec![Schwarz],
        vec![Schwarz, Gelb],
        vec![Schwarz],
        vec![Rot, Tuerkis],
        vec![Blau, Schwarz],
        vec![Gelb],
        vec![Tuerkis, Gelb],
        vec![Tuerkis],
        vec![Gelb],
    ];
    defs.into_iter()
        .enumerate()
        .map(|(i, colors)| BonusChip { chip_id: i, colors })
        .collect()
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::tile::TileColor::*;

    #[test]
    fn pools_have_expected_sizes() {
        let pool = build_dome_tile_pool();
        assert_eq!(pool.len(), 18);
        assert!(pool.iter().all(|t| t.spaces.len() == 4));
        // tile_ids 0..18 fortlaufend
        for (i, t) in pool.iter().enumerate() {
            assert_eq!(t.tile_id, i);
        }
        assert_eq!(build_bonus_chip_pool().len(), 20);
    }

    #[test]
    fn space_accepts_logic() {
        let normal = DomeSpace::normal(Rot);
        assert!(normal.accepts(Rot));
        assert!(!normal.accepts(Blau));
        assert!(DomeSpace::wild().accepts(Schwarz));
        // Special akzeptiert keinen normalen Stein, locked
        let sp = DomeSpace::special();
        assert!(!sp.accepts(Rot));
        assert!(!sp.accepts_special()); // weil is_locked
    }

    #[test]
    fn rotation_preserves_set_and_layout() {
        // 180° zweimal == original
        let mut t = DomeTile::new(
            0,
            vec![
                DomeSpace::normal(Blau),
                DomeSpace::normal(Gelb),
                DomeSpace::normal(Rot),
                DomeSpace::normal(Schwarz),
            ],
            0,
        );
        let orig = t.spaces.clone();
        t.apply_rotation(180).unwrap();
        assert_eq!(t.spaces[0].required_color, Some(Schwarz));
        t.apply_rotation(180).unwrap();
        assert_eq!(t.spaces, orig);
    }

    #[test]
    fn special_unlocks_when_others_filled() {
        // Platte mit 1 Special + 3 Normal; fülle die 3 Normal → Special unlockt.
        let mut t = build_dome_tile_pool()[0].clone(); // [Gelb, Schwarz, Tuerkis, Special]
        let sp_idx = t.special_space_idx().unwrap();
        for i in 0..4 {
            if i != sp_idx {
                let col = t.spaces[i].required_color.unwrap();
                t.spaces[i].placed_color = Some(col);
            }
        }
        assert!(t.try_unlock_special());
        assert!(t.spaces[sp_idx].accepts_special());
        // zweiter Aufruf liefert false (schon entsperrt)
        assert!(!t.try_unlock_special());
    }
}
